tests/software_tests/database/data_record/test_raw_data_record.py

```diff
@@ -0,0 +1,155 @@
+        mock_isinstance.assert_called_once_with(value, int)
+
+    @pytest.mark.parametrize(
+        "value", [0, -1]
+    )
+    def test_length_setter_value_error(self, value):
+        with pytest.raises(ValueError):
+            RawDataRecord.length.fset(self.mock_data_record, value)
+
+    # max_raw_value
+
+    @pytest.mark.parametrize(
+        "length, value", [
+            (2, 3),
+            (5, 31),
+            (8, 255),
+        ]
+    )
+    def test_max_raw_value_getter(self, length, value):
+        raw_data_record = RawDataRecord("TestRawDataRecord", length)
+        assert raw_data_record.max_raw_value == value
+
+    # is_reoccurring
+
+    def test_is_reoccurring_getter(self):
+        assert RawDataRecord.is_reoccurring.fget(self.mock_data_record) is False
+
+    # min_occurrences
+
+    def test_min_occurrences_getter(self):
+        assert RawDataRecord.min_occurrences.fget(self.mock_data_record) == 1
+
+    # max_occurrences
+
+    def test_max_occurrences_getter(self):
+        assert RawDataRecord.max_occurrences.fget(self.mock_data_record) == 1
+
+    # contains
+
+    def test_contains_getter(self):
+        assert RawDataRecord.contains.fget(self.mock_data_record) == ()
+
+    # decode
+
+    @pytest.mark.parametrize(
+        "value", [1, 4]
+    )
+    @patch(f"{SCRIPT_LOCATION}.DecodedDataRecord")
+    def test_decode(self, mock_decoded_data_record, value):
+        self.mock_data_record.max_raw_value = 8
+        assert RawDataRecord.decode(self.mock_data_record, value) == mock_decoded_data_record.return_value
+        mock_decoded_data_record.assert_called_once_with(
+            name=self.mock_data_record.name,
+            raw_value=value,
+            physical_value=value
+        )
+
+    @pytest.mark.parametrize(
+        "value", ["test", None]
+    )
+    @patch(f"{SCRIPT_LOCATION}.isinstance")
+    def test_decode_type_error(self, mock_isinstance, value):
+        with pytest.raises(TypeError):
+            RawDataRecord.decode(self.mock_data_record, value)
+        mock_isinstance.assert_called_once_with(value, int)
+
+    @pytest.mark.parametrize(
+        "value, max_raw_value", [
+            (-1, 2),
+            (3, 2),
+            (16, 6),
+        ]
+    )
+    def test_decode_value_error(self, value, max_raw_value):
+        self.mock_data_record.max_raw_value = max_raw_value
+        with pytest.raises(ValueError):
+            RawDataRecord.decode(self.mock_data_record, value)
+
+    # encode
+
+    @pytest.mark.parametrize(
+        "value", [1, 4]
+    )
+    def test_encode(self, value):
+        self.mock_data_record.max_raw_value = 8
+        assert RawDataRecord.encode(self.mock_data_record, value) == value
+
+    @pytest.mark.parametrize(
+        "value", ["test", None]
+    )
+    @patch(f"{SCRIPT_LOCATION}.isinstance")
+    def test_encode_type_error(self, mock_isinstance, value):
+        with pytest.raises(TypeError):
```

Comment:
we need one more line before with pytest.raises:
mock_isinstance.return_value = False to make sure that type check failed